### PR TITLE
Refactor schema validator code

### PR DESCRIPTION
**Pull request by faisal-amin82**
Currently schema validation is performed for every request. Using
sync package's Once struct makes sure the code execute only once.
This operation is also safe from concurrency point of view.
We also refactored main function where instead of creating a service
layer and passing it to register funciton, we initialize the service
struct directly in Register function.

Resovles: DSND-122

### CODE REVIEW

File: validation/chOpenApiValidator.go [hunk 1]
```diff
@@ -246,3 +233,37 @@ func handleParseError(pe *openapi3filter.ParseError) models.CHError {
 func findRoute(r routers.Router, req *http.Request) (route *routers.Route, pathParams map[string]string, err error) {
 	return r.FindRoute(req)
 }
+
+func loadSchemaFromFile(ctx context.Context, openApiSpec, contextId string) (*openapi3.T, error){
```
Comment (mreis-ch) [inline validation/chOpenApiValidator.go:237]: Just from an ordering, maybe move this to after the getSchema function

File: validation/chOpenApiValidator.go [hunk 2]
```diff
@@ -246,3 +233,37 @@ func handleParseError(pe *openapi3filter.ParseError) models.CHError {
...
+	log.Info("This will execute only once")
```
Comment (mreis-ch) [inline validation/chOpenApiValidator.go:238]: Don't need this log message

File: validation/chOpenApiValidator.go [hunk 3]
```diff
@@ -30,93 +31,80 @@ var (
...
+func (chv *CHValidatorImpl) ValidateRequestAgainstOpenApiSpec(httpReq *http.Request, openApiSpec, contextId string) ([]byte, error) {
 
 	// Get the Open API 3 validation schema location.
 	ctx := context.Background()
-	loader := &openapi3.Loader{Context: ctx, IsExternalRefsAllowed: true}
-	abs, err := callFilepathAbs(openApiSpec)
-	if err != nil {
-		log.ErrorC(contextId, err, log.Data{config.MessageKey: "error occurred while retrieving absolute path of validation schema file"})
+
+	if err := chv.GetSchema(ctx, openApiSpec, contextId); err != nil {
```
Comment (amoriartyCH) [inline validation/chOpenApiValidator.go:61]: Does this have to be exported, do we use it anywhere outside of the package? (GetSchema)

File: validation/chOpenApiValidator.go [hunk 4]
```diff
@@ -246,3 +234,33 @@ func handleParseError(pe *openapi3filter.ParseError) models.CHError {
 func findRoute(r routers.Router, req *http.Request) (route *routers.Route, pathParams map[string]string, err error) {
 	return r.FindRoute(req)
 }
+
+func (chv *CHValidatorImpl) GetSchema(ctx context.Context, openApiSpec, contextId string) error {
```
Comment (amoriartyCH) [inline validation/chOpenApiValidator.go:238]: As per previous comment, if this does need to be exported we will need to comment it's behaviour

File: validation/chOpenApiValidator.go [hunk 5]
```diff
@@ -246,3 +234,33 @@ func handleParseError(pe *openapi3filter.ParseError) models.CHError {
...
+func loadSchemaFromFile(ctx context.Context, openApiSpec, contextId string) (*openapi3.T, error){
+	loader := &openapi3.Loader{Context: ctx, IsExternalRefsAllowed: true}
+	abs, err := callFilepathAbs(openApiSpec)
+	if err != nil {
+		log.ErrorC(contextId, err, log.Data{config.MessageKey: "error occurred while retrieving absolute path of validation schema file"})
+		return nil, err
+	}
+	log.InfoC(contextId, fmt.Sprintf("Retrieved absolute path of validation schema "), log.Data{config.SchemaAbsolutePathKey: abs})
+
+	// Load the validation schema.
```
Comment (amoriartyCH) [inline validation/chOpenApiValidator.go:264]: Slightly outdated comment here, maybe Return schema?

Could also comment some other lines of code within this function, to make it clean what we're doing at each step?

File: validation/chOpenApiValidator_test.go [hunk 6]
```diff
@@ -56,9 +58,9 @@ func TestValidateRequestAgainstOpenApiSpecFailsAbs(t *testing.T) {
 // TestValidateRequestAgainstOpenApiSpecFailsFileOpen asserts that when calling to load the file using an ABS path fails
 // it is handled correctly.
 func TestValidateRequestAgainstOpenApiSpecFailsFileOpen(t *testing.T) {
+	callOnce = sync.Once{}
```
Comment (amoriartyCH) [inline validation/chOpenApiValidator_test.go:61]: What does adding a new callOnce every time do for these tests? Is it worth putting a small comment above it just to clarify why we re-create the once? guessing to do with the once == true after first test is run so all others fail.

File: validation/chOpenApiValidator.go (hunk 1 above)
Comment (faisal-amin82) [reply validation/chOpenApiValidator.go:254]: Reordered as Matt suggested

File: validation/chOpenApiValidator.go (hunk 2 above)
Comment (faisal-amin82) [reply validation/chOpenApiValidator.go]: Removed the log line

File: validation/chOpenApiValidator.go (hunk 3 above)
Comment (faisal-amin82) [reply validation/chOpenApiValidator.go]: Reworked this logic to make getSchema as non exported function as this is only used in the same package

File: validation/chOpenApiValidator.go (hunk 4 above)
Comment (faisal-amin82) [reply validation/chOpenApiValidator.go]: Same as above

File: validation/chOpenApiValidator_test.go (hunk 6 above)
Comment (faisal-amin82) [reply validation/chOpenApiValidator_test.go:61]: Yes exactly as you said Aaron, in order to test all these unit tests collectively, we need to reset once as this causes the `doc` to not get initialised after subsequent calls

File: validation/chOpenApiValidator.go [hunk 7]
```diff
@@ -246,3 +235,33 @@ func handleParseError(pe *openapi3filter.ParseError) models.CHError {
 func findRoute(r routers.Router, req *http.Request) (route *routers.Route, pathParams map[string]string, err error) {
 	return r.FindRoute(req)
 }
+
+func getSchema(ctx context.Context, openApiSpec, contextId string, chv *CHValidatorImpl) (err error) {
+	//callOnce.Do method insures the code is executed once so that the schema is not repeatedly loaded unnecessarily.
```
Comment (amoriartyCH) [inline validation/chOpenApiValidator.go:240]: Super picky: I believe insures should be ensures

**Review by ch-tstrickland — APPROVED** [review]

**Review by mreis-ch — APPROVED** [review]

**Pull request merged**